code/android-core-library/src/phone/java/com/adobe/marketing/mobile/services/caching/FileCacheService.java

```diff
@@ -0,0 +1,126 @@
+
+    private final CacheFileManager cacheFileManager;
+
+    public FileCacheService(@NonNull final DeviceInforming deviceInfoService) {
+        this.cacheFileManager = new CacheFileManager(deviceInfoService, ROOT_CACHE_DIR_NAME);
+    }
+
+    /**
+     * Creates or updates the key-value pair in the cache.
+     * Stores the content and its metadata as separate files internally in the
+     * cache directory supplied by {@code DeviceInforming#getApplicationCacheDir}
+     *
+     * @param cacheName name of the bucket where the cache entry is to be created
+     * @param key       key for the cache entry
+     * @param value     the value that is to be associated with {@code key}
+     * @return true if the value for the key was created or updated; false otherwise.
+     */
+    @Override
+    public boolean set(@NonNull String cacheName, @NonNull String key, @NonNull CacheEntry value) {
+        if (!canProcess(cacheName, key)) return false;
+
+        // Create the bucket if necessary
+        final File cacheBucket = cacheFileManager.createCache(cacheName);
+        if (cacheBucket == null) {
+            Log.debug(ServiceConstants.LOG_TAG, TAG, "Could not set value for key: [%s] in cache: [%s]." +
+                    "Cache creation failed.");
+            return false;
+        }
+
+        return cacheFileManager.createCacheFile(cacheName, key, value);
+    }
+
+    /**
+     * Retrieves the value associated with the key being queried.
+     * Note that the {@code CacheResult} returned may have additional metadata
+     * internally used by the {@code FileCacheService}.
+     *
+     * @param cacheName name of the bucket where the entry is to be fetched from
+     * @param key       key for the cache entry
+     * @return the {@code FileCacheResult} associated with the key if present;
+     * null otherwise.
+     */
+    @Nullable
+    @Override
+    public CacheResult get(@NonNull String cacheName, @NonNull String key) {
+        if (!canProcess(cacheName, key)) return null;
+
+        final File cacheFile = cacheFileManager.getCacheFile(cacheName, key);
+
+        if (cacheFile == null) return null;
+
+        final Map<String, String> cacheMetadata = cacheFileManager.getCacheMetadata(cacheName, key);
+
+        // We lost the metadata/metadata is corrupt, but have an entry - so there must have been an error.
+        // Remove the cache entry as well and return null.
+        if (cacheMetadata == null) {
+            Log.debug(ServiceConstants.LOG_TAG, TAG, "Could not find metadata for key: [%s] in cache: [%s].");
+            remove(cacheName, key);
+            return null;
+        }
+
+        final String expiryFromMetadata = cacheMetadata.get(FileCacheResult.METADATA_KEY_EXPIRY_IN_MILLIS);
+        final CacheExpiry expiry = (expiryFromMetadata == null)
+                ? CacheExpiry.never()
+                : CacheExpiry.at(new Date(Long.parseLong(expiryFromMetadata)));
```

Comment:
`Long.parseLong` could throw an exception if the file is corrupted. Can you handle that case as well?